### ISSUE TITLE
Implement Basic App Platform

### ISSUE THREAD

**Issue by ecton** [khonsulabs/bonsaidb#55]
The goal of this issue is to create a basic app platform that allows defining restful API endpoints, simple HTTP serving, a custom over-the-wire (WebSocket/PliantDb) API. A developer could use this platform to create an app that uses PliantDb to manage users and permissions, and optionally an HTTP layer with WebSocket support serving requests. The HTTP layer's main purposes are writing RESTful APIs and serving a single-page app that is powered by WASM + PliantDb over WebSockets.

* [x] Implement custom api support (#54)
* [x] Add Users and Permissions management
  * [x] Ability to define permission groups
  * [x] Ability to define roles (one or more groups)
  * [x] Ability to create users (username + password minimum)
  * [x] Ability to assign users to roles
* [x] Add connected-client tracking, a-la basws:
  * [x] Upon connecting, initiate authentication
  * [x] If no auth, proceed with default permissions
    * [x] If not permitted to connect, disconnect.
  * [x] If authed, load user profile and effective permissions (needs ~khonsulabs/actionable#2~).
  * [x] Call Backend method notifying of either anonymous or logged-in connection.
* [x] Implement HTTP layer
  * [ ] Simple routing -> handler mapping. Maybe doable with actionable + another trait?
  * [ ] Handlers convert from HttpRequest to HttpResponse
  * [x] Integrate WebSockets into routing
  * [x] Add TLS support
  * [x] Add static-file handler(s)
  * [ ] Add Single-Page-App server

**Comment by ecton** [comment]
I was investigating adding optional HTTP support via `routerify` and replace the server's websockets with it. It looks like `routerify` is going through an update to 2.0, and it might take the ecosystem a little while to get caught up. Currently `routerify-websocket` introduces old tokio dependencies, which I'd prefer to avoid. I want to wait until it is updated. It appears work is underway.

Overall, I like the design of `routerify`'s API, so I'd rather wait and adopt it than seek other solutions. So, for now I'm going to just table this project for a little while and focus on other areas.

**Comment by ecton** [comment]
Tokio has a new project that looks very tempting: https://tokio.rs/blog/2021-07-announcing-axum

It honestly feels like a better choice as I can investigate exposing the HTTP layer as a tower service? Not really sure, I need to spend time understanding how tower combines services, but it seems like this is a good direction to explore.

**Comment by ecton** [comment]
In #97, I've implemented this by extending the `Backend` trait with an extra method when the `http` feature flag is enabled (`server-http` for the omnibus crate). I've added an example that shows how to use this to serve requests using the [`Axum`](https://github.com/tokio-rs/axum) crate. Additionally, #95 added automatic ACME certificate acquisition.

This issue as it stands has been implemented, although we still have more a grandiose vision for the platform as a whole.
